daffodil-lib/src/main/scala/edu/illinois/ncsa/daffodil/util/DecimalUtils.scala

```diff
@@ -0,0 +1,342 @@
+      Integer.parseInt(String.valueOf('0'), 16))
+
+    val positives: List[Int] = List(Integer.parseInt(String.valueOf(chars(0)), 16))
+    val negatives: List[Int] = List(Integer.parseInt(String.valueOf(chars(1)), 16))
+    val unsigneds: List[Int] = List(Integer.parseInt(String.valueOf(chars(2)), 16))
+    val zero_signs: List[Int] = List(Integer.parseInt(String.valueOf(chars(3)), 16))
+
+    val hexCodes = policy match {
+      case BinaryNumberCheckPolicy.Strict => Map("positive" -> positives, "negative" -> negatives, "unsigned" -> unsigneds, "zero_sign" -> zero_signs)
+      case BinaryNumberCheckPolicy.Lax => Map("positive" -> valid_positives, "negative" -> valid_negatives, "unsigned" -> valid_unsigneds, "zero_sign" -> valid_zero_signs)
+    }
+
+    hexCodes
+  }
+
+  def packedToBigInteger(num: Array[Byte], signCodes: String, policy: BinaryNumberCheckPolicy): JBigInteger = {
+    val hexCodes: Map[String, List[Int]] = signCodesToHex(signCodes, policy)
+    val numDigits: Int = num.size * 2  // 2 digits stored per byte
+    val outputData: Array[Char] = new Array[Char](numDigits-1)
+    var outputPos: Int = 0
+    var offset: Int = 0
+    var nibble: Int = 0
+    var negative: Boolean = false
+
+    // Parse and validate the last (sign) bit
+    nibble = (num(offset +  num.size - 1) & 0x0F)
+    if (hexCodes("negative").contains(nibble)) {
+      negative = true
+    } else if (!hexCodes("positive").contains(nibble)) {
+      throw new NumberFormatException("Invalid low nibble")
+    }
+
+    while (outputPos < outputData.size - 1) {
+      // Parse high nibble
+      nibble = (num(offset) & 0xFF) >>> 4
+      if (nibble > 0x09) {
+        throw new NumberFormatException("Invalid high nibble")
+      }
+
+      outputData(outputPos) = (nibble | 0x0030).toChar
+      outputPos = outputPos + 1
+
+      // Parse low nibble
+      nibble = (num(offset) & 0x0F)
+      offset = offset + 1
+      if (nibble > 0x09) {
+        throw new NumberFormatException("Invalid low nibble")
+      }
+
+      outputData(outputPos) = (nibble | 0x0030).toChar
+      outputPos = outputPos + 1
+    }
+
+    // Parse last digit
+    nibble = (num(offset) & 0xFF) >>> 4
+    if (nibble > 0x09) {
+      throw new NumberFormatException("Invalid high nibble")
+    }
+
+    outputData(outputPos) = (nibble | 0x0030).toChar
+
+    if (negative)
+      new JBigInteger(new String(outputData)).negate()
+    else
+      new JBigInteger(new String(outputData))
+
+  }
+
+  def packedToBigDecimal(num: Array[Byte], scale: Int, signCodes: String, policy: BinaryNumberCheckPolicy): JBigDecimal = {
+    return new JBigDecimal(packedToBigInteger(num, signCodes, policy), scale)
+  }
+
+  def packedFromBigInteger(num: String, nBits: Int, signCodes:String, policy: BinaryNumberCheckPolicy): Array[Byte] = {
+    val negative: Boolean = (num.charAt(0) == '-')
+    // Discard any sign symbols or decimal points from the string (decimal point will be maintained in binaryVirtualDecimalPoint)
```

Comment:
It might make the most sense to just remove the part about binaryVirtualDecimalPoint from the comment.  I was trying to convey that when we are creating the packed/bcd/ibm4690 from the BigInteger string that we do not need to keep track of the decimal point as the unparser already has the information on where the decimal point is supposed to go in the binaryVirtualDecimalPoint property.

I think I'll go ahead and change the function name to _FromBigNumber as this function works on both BigInteger numbers and BigDecimal numbers that have been converted to string.

Reply:
I'm not even sure dfdl:binaryDecimalVirtualPoint works for unparsing, or maybe I just don't understand how it works. The way it works in parsing is you parse a number as an Int and then shift the decimal point binaryDeciamlVirtualPoint number of digist. On unparsing, we start with a decimal, so does that mean we should shift the decimal the opposite of binaryDecimalVirtualPoint to give an integer, then we just unparse the integer? And it should be an SDE if after the shift we don't get an integer? The spec isn't very clear about this. And it makes me think that this function should just assume/assert that there are no decimal points and that the number is always an int?

Reply:
I may not have worded things correctly.  I think we are both trying to make the same point, that when both parsing and unparsing we are essentially only dealing with integer values in these byte arrays.  In the case of parsing, we shift the decimal point around based on whats in BVDP.  When unparsing BVDP isn't used, but I just wanted to convey the fact that even though I am stripping the decimal point from the input string for the FromBigInteger functions, the decimal point information isn't being lost as it is still in BVDP in the schema.

Reply:
But I think BVDP *should* be used during unparsing somewhere. For example the BinaryDecimalUnparserBase scales the BigDecimal by BVDP and then converts it to an BigInt and unparses the BigInt. I think we need to do something similar for these packed things. They should take the Decimal from the infost, scale it by BVDP, convert it to a BigInt, and then unparse that. (Note BinaryDecimalUnparseBase does not appear to do any error checking to ensure the BigDecimal converts to a BigInt after scalling, might be a bug). Does what I describe happen anywhere for the packed stuff?

Reply:
Yes, it is symmetric. BinaryDecimalVirtualPoint is for what is called Fixed-point decimal. It provides a scale factor that converts an integer to a decimal when parsing, and a decimal back to an integer when unparsing. Data should round-trip. This is used all the time in financial data where floating point is never used, base 10 rounding rules are required, etc. 

There's an obscure case, that *is* used I believe which is when the binaryDecimalVirtualPoint is negative when it scales an integer (on parsing) into a bigger integer, and cuts it down to a smaller integer on unparsing (have to think about rounding in this latter case, as the digits being lost aren't necessarily zeros). 

This is used for currency units like Japanese Yen, where often "000" are removed, so that what is stored in the data is thousands of Yen, but logically the data is single Yen. (Yen are like pennies - so thousands of yen is like $10 units, roughly in concept.)

Reply:
Ok, I understand what you are gettting at now.  Currently what I am doing is not converting the BigDecimal to a BigInteger, instead I am just calling bigDec.toPlainString and doing fromBigInteger on that string, and then in FromBigInteger I am removing the decimal point character from the string.

What you are suggesting would probably be clearer in the code and removes the need for me to run a regex on the input string in FromBigInteger.  I hadn't really considered a malformed infoset when unparsing.  I'll start working on these changes.

Reply:
Yeah, what you're doing is very similar. But it's possible that extra zeros could be added based on the BDVL or rounding issues as Mike mentions. That has to be dealt with somehow.

Reply:
Section 13.7.1 says when unparsing a binary number (packed decimal or twos-complement) and excess precision is supplied in the Infoset no rounding occurs. It is a processing error.

This to me implies that if the infoset element is scaled based on BDVP and the result is not an integer, then it is a processing error.